lib/index.ts

```diff
@@ -0,0 +1,572 @@
+				{ reqType: 'GET', path: 'conversations/{conversation_id}/inboxes' }),
+			listMessages: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'conversations/{conversation_id}/messages[page={page}][limit={limit}]' }),
+			update: _.partial(this.httpCall,
+				{ reqType: 'PATCH', path: 'conversations/{conversation_id}' }),
+		};
+		this.inbox = {
+			create: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'inboxes' }),
+			get: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes/{inbox_id}' }),
+			list: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes' }, null),
+			listChannels: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes/{inbox_id}/channels' }),
+			listConversations: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes/{inbox_id}/conversations[q{q}...][page={page}][limit={limit}]' }),
+			listTeammates: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes/{inbox_id}/teammates' }),
+		};
+		this.message = {
+			get: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'messages/{message_id}' }),
+			receiveCustom: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'channels/{channel_id}/incoming_messages' }),
+			reply: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'conversations/{conversation_id}/messages' }),
+			send: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'channels/{channel_id}/messages' }),
+		};
+	}
+
+	private httpCall = (details: FrontRequest, params: any, callback?: InternalCallback): Promise<any | void> => {
+		const requestOpts = {
+			body: (params) ? params : {},
+			headers: {
+				Authorization: `Bearer ${this.apiKey}`
+			},
+			json: true,
+			method: details.reqType,
+			url: `${FrontURL}/${this.formatPath(details.path, params)}`
+		};
+
+		// Make the request.
+		return request(requestOpts).promise().catch((error: any) => {
+			// Format this into something useful.
+			throw error.error._error;
+		}).asCallback(callback);
+	}
+
+	private formatPath(path: string, data: FrontRequestData | null): string {
+		const replaceTag = (substring: string) => {
+			// Test to see if this is an array type.
+			// If so, we need to repeat all entries in the array.
+			const preamble = substring.substring(1, substring.indexOf('{'));
+			let newString = substring;
+			let arraySubstitution = false;
+			if (substring.indexOf('...') !== -1) {
+				newString = _.replace(newString, /\.\.\./g, '');
+				arraySubstitution = true;
+			}
+			return _.replace(newString, /\{.*\}/g, (candidate: string) => {
+				// Search for the replacee in the data.
+				const replaceTrimmed = _.trimEnd(_.trim(candidate, '{'), '}');
+				if (testData[replaceTrimmed]) {
+					if (!arraySubstitution) {
+						return testData[replaceTrimmed];
+					} else {
+						let arrayString = '';
+						if (testData[replaceTrimmed].length === 0) {
+							return;
+						}
+
+						for (let index = 0; index < testData[replaceTrimmed].length; index += 1) {
```

Comment:
Personally I'm used to seeing `++` in `for` loops, although it's probably nicer to just use
```typescript
testData[replaceTrimmed].map(function(item, index) {
	...
})